reps/2023-12-04-accelerated-dag.md

```diff
@@ -0,0 +1,360 @@
+out-of-band, as part of the application code. However, to realize the
+long-term goals, we will introduce an API that captures collective
+communication ops as part of the DAG. An example API:
+
+```python
+workers = [Worker.remote() for worker in workers]
+# A task called on the CollectiveGroup will be gang-scheduled on each worker.
+pool = CollectiveGroup(workers)
+outputs = pool.allreduce.bind(inputs)
+```
+
+Workloads
+=========
+
+Possible validation workloads fall under two different categories:
+
+**Common patterns in distributed inference and training:** These are
+workloads that may be distributed but only use one type of parallelism
+(e.g., tensor- vs pipeline-parallel) and only one model. They are
+typically already supported by other multi-controller frameworks such as
+DeepSpeed, Distributed TF/PyTorch, etc.
+
+Goals:
+
+-   Performance parity to validate system overheads
+-   Validate API expressivity - we want to show that it is much simpler
+    to implement these patterns with a single-controller interface
+-   Increase the flexibility of (some) patterns. For example, pipeline
+    parallelism is a well-known technique provided by Deepspeed, but
+    Deepspeed only supports it for sequential
+    modules
+    ([link](https://www.deepspeed.ai/tutorials/pipeline/#getting-starting-with-pipeline-parallelism)),
+    meaning that it can't be used for DAGs of modules.
+
+**Advanced patterns:** These workloads may use multiple types of
+parallelism, multiple models, and/or are more "dynamic" in some sense
+(e.g., providing low down time after cluster size changes). In many
+cases, there may be no existing popular implementation.
+
+Goals:
+
+-   Validate API expressivity
+-   Validate integration with higher-level ML frameworks
+-   Enable new workloads via Ray's superior flexibility.
+
+We plan to implement all of the common patterns during prototype
+validation, and will select some advanced patterns as we are closer to
+release. Feedback on workload selection is welcome! The following
+categories are in rough order of technical complexity:
+
+### Common patterns: Microbenchmarks (simple distributed inference)
+
+|                                                                 | Key properties / requirements                                        | Goals                                                            |
+|-----------------------------------------------------------------|----------------------------------------------------------------------|------------------------------------------------------------------|
+| vLLM tensor parallelism                                         |                                                                      | Reduce Ray overheads                                             |
+| vLLM pipeline parallelism                                       | P2P/cross-node GPU communication                                     | Reduce (expected) Ray overheads; Validate cross-node performance  |
```

Comment:
My understanding from vLLM folks' benchmarks was that cross-node GPU communication overhead is too high (even if using optimized transports) so for now they don't have a desire to use it. I could be wrong though, this is hearsay.

Reply:
For tensor parallelism, yes, but I don't think that's the case for pipeline parallelism.